src/library_manager/lib_manager.c

```diff
@@ -492,50 +577,74 @@ static struct comp_dev *lib_manager_module_create(const struct comp_driver *drv,
 						  const struct comp_ipc_config *config,
 						  const void *spec)
 {
+	const struct sof_man_fw_desc *const desc = lib_manager_get_library_manifest(config->id);
 	const struct ipc_config_process *args = (struct ipc_config_process *)spec;
-	const uint32_t module_id = IPC4_MOD_ID(config->id);
-	const uint32_t instance_id = IPC4_INST_ID(config->id);
-	const uint32_t log_handle = (uint32_t)drv->tctx;
-	const struct module_interface *ops;
-	byte_array_t mod_cfg;
+	const uint32_t entry_index = LIB_MANAGER_GET_MODULE_INDEX(config->id);
+	const struct module_interface *ops = NULL;
+	const struct sof_man_module *mod;
+	system_agent_start_fn agent;
+	void *adapter_priv = NULL;
+	void **agent_iface;
+	struct comp_dev *dev;
 	int ret;
 
-	/*
-	 * Variable used by llext_manager to temporary store llext handle before creation
-	 * a instance of processing_module.
-	 */
-	struct comp_dev *dev;
+	tr_dbg(&lib_manager_tr, "start");
+	if (!desc) {
+		tr_err(&lib_manager_tr, "Error: Couldn't find loadable module with id %u.",
+		       config->id);
+		return NULL;
+	}
 
-	/* At this point module resources are allocated and it is moved to L2 memory. */
-	const uint32_t module_entry_point = lib_manager_allocate_module(config,
-									args->data);
+	if (entry_index >= desc->header.num_module_entries) {
+		tr_err(&lib_manager_tr, "Entry index %u out of bounds.", entry_index);
+		return NULL;
+	}
+
+	mod = (const struct sof_man_module *)
+		((const uint8_t *)desc + SOF_MAN_MODULE_OFFSET(entry_index));
+
+	switch (lib_manager_get_module_type(desc, mod)) {
+	case MOD_TYPE_LLEXT:
+	case MOD_TYPE_LMDK:
+		agent = &native_system_agent_start;
+		agent_iface = (void **)&ops;
```

Comment:
`(const void **)`

Reply:
The goal of these changes is unification of the interface for `native_system_agent_start` and `system_agent_start`. The problem is that `system_agent_start`, in an `iface`, doesn't return a pointer to the `const module interface`, but rather an adapter object, which is not const. Adding const here will either force a type cast elsewhere or artificially split code paths and create code duplication.

Reply:
@softwarecki but you can change `system_agent_start()` too - that's an SOF internal function, not a part of an external API?